### ISSUE TITLE
[BUG] Windows CI jobs are very slow

### ISSUE THREAD

**Issue by scoder** [cython/cython#4963]
I noticed that the Windows build jobs take very long to finish, e.g. ~40 minutes (Windows) vs. ~9 minutes (Linux). This is not a hardware issue, the VMs seem to be much faster according to the end-of-run timings (e.g. Cython translation takes 150s on Windows vs. 440s on Linux). But the C compilation times are completely the opposite, e.g. 10,000s on Windows vs. 220s on Linux.

Probable reasons are:
1) missing ccache
2) different C compiler options

We are clearly wasting computing resources here, and also make us wait a long time to get CI feedback on changes. Any improvements would be welcome.

**Comment by 0dminnimda** [comment]
At the time I tried to serarch for compiler options to make compilation faster but I couldn't get them to work as easily so I abounded the idea as it wasn't a requirement.

**Comment by matusvalo** [comment]
One easy fix is that full CI is skiped for documentation-only changes. E.g. we can specify to run tests only when files in `Cython/*` are changed [1].

E.g. pylint is doing something simillar [2]


[1] https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore
[2] https://github.com/PyCQA/pylint/blob/e0ecc35e93c4dc32b691cfb8a3ad526ebfafb086/.github/workflows/primer-test.yaml#L1-L13

**Comment by scoder** [comment]
> full CI is skiped for documentation-only changes

Yes, that's also a good idea. We'd need a separate documentation build workflow for that, though. Currently, everything is done in the same bash script and the same matrix config. However, I think a separate workflow for documentation makes sense in any case. It's really different enough. Splitting out a separate workflow config for it shouldn't be difficult, it can still use the same bash script.

**Comment by scoder** [comment]
> At the time I searched for compiler options to make compilation faster but I couldn't get them to work as easily so I abounded the idea as it wasn't a requirement.

I think the most worthwhile is some kind of ccache-like caching mechanism. Since many test C files are not impacted by most changes to Cython, completely avoiding to recompile them makes a *huge* difference. (I think the macos builds also still lack ccache.)

**Comment by 0dminnimda** [comment]
When CYTHON_RUNTESTS_CCACHE is set?
https://github.com/cython/cython/blob/278d07603e9ae9ce167f489c3c9855cb424cd5e7/runtests.py#L504

**Comment by 0dminnimda** [comment]
Also, I cannot evluate how successful ccache was for windows as windows run is failing on building cython

**Comment by scoder** [comment]
> When CYTHON_RUNTESTS_CCACHE is set?

I don't think it's used anywhere. The ccache setup for our test runs happens in ci-run,sh.

**Comment by scoder** [comment]
I fixed some of the Windows failures in master. Only some walrus error tests are still failing now, apparently because of a missing exception message on a SyntaxError.

**Comment by da-woods** [comment]
> Only some walrus error tests are still failing now, apparently because of a missing exception message on a SyntaxError.

That's most likely that the `compile` replacement function isn't capturing the output successfully on Windows. I'd be very surprised if the parser isn't emitting the same SyntaxError on Windows.
